python_modules/libraries/dagster-aws/dagster_aws/externals.py

```diff
@@ -0,0 +1,29 @@
+import random
+import string
+from typing import Optional
+
+import boto3
+from botocore.exceptions import ClientError
+from dagster._core.external_execution.resource import (
+    ExternalExecutionParams,
+)
+from dagster._core.external_execution.utils import ExternalExecutionBlobStoreMessageReader
+
+
+class ExternalExecutionS3MessageReader(ExternalExecutionBlobStoreMessageReader):
+    def __init__(self, *, interval: int = 10, bucket: str, client: boto3.client):
+        super().__init__(interval=interval)
+        self.bucket = bucket
+        self.key_prefix = "".join(random.choices(string.ascii_letters, k=30))
+        self.client = client
+
+    def get_params(self) -> ExternalExecutionParams:
+        return {"bucket": self.bucket, "key_prefix": self.key_prefix}
+
+    def download_messages_chunk(self, index: int) -> Optional[str]:
```

Comment:
how is the index state tracked by readers, if a reader process restarts will it lose this state?

Reply:
>how is the index state tracked by readers

Both the reader and writer just start a counter at 1 and increment when they perform a successful read/write respectively.

> if a reader process restarts will it lose this state

Yes, I haven't been thinking about restarting the reader process, since this is the op execution process, and if it restarts it will launch a new external process.